File: sdk-py/paramant/wire_format.py
"""
Paramant wire format v1 (PQHB) — Python encoder / decoder.

Implements the spec in docs/wire-format-v1.md byte-for-byte. The relay's
Node implementation in relay/crypto/wire-format.js is authoritative; this
module MUST produce blobs that decode cleanly there.

Public API:
    encode(kem_id, sig_id, ct_kem, sender_pub, signature, nonce, ciphertext) -> bytes
    decode(blob) -> dict
    build_aad(kem_id, sig_id, flags=0, chunk_index=0) -> bytes
    is_v1(blob) -> bool
"""
from __future__ import annotations

from typing import Optional

from .errors import InvalidFlags, InvalidMagic, InvalidVersion, MalformedBlob

MAGIC = b"PQHB"
VERSION = 1
SUPPORTED_VERSIONS = (VERSION,)
HEADER_SIZE = 10
NONCE_SIZE = 12
_U16_MAX = 0xFFFF
_U32_MAX = 0xFFFFFFFF


def _u16_be(v: int) -> bytes:
    if not (0 <= v <= _U16_MAX):
        raise ValueError(f"value {v} out of range for uint16")
    return v.to_bytes(2, "big")


def _u32_be(v: int) -> bytes:
    if not (0 <= v <= _U32_MAX):
        raise ValueError(f"value {v} out of range for uint32")
    return v.to_bytes(4, "big")


def encode(
    kem_id: int,
    sig_id: int,
    ct_kem: bytes,
    sender_pub: bytes,
    signature: Optional[bytes],
    nonce: bytes,
    ciphertext: bytes,
    flags: int = 0x00,
) -> bytes:
    """Produce a v1 blob. See docs/wire-format-v1.md."""
    if not (0 <= kem_id <= _U16_MAX):
        raise ValueError("kem_id must be uint16")
    if not (0 <= sig_id <= _U16_MAX):
        raise ValueError("sig_id must be uint16")
    if flags != 0x00:
        raise InvalidFlags(flags)
    if not isinstance(ct_kem, (bytes, bytearray)):
        raise TypeError("ct_kem must be bytes")
    if not isinstance(sender_pub, (bytes, bytearray)):
        raise TypeError("sender_pub must be bytes")
    if not isinstance(nonce, (bytes, bytearray)) or len(nonce) != NONCE_SIZE:
        raise ValueError(f"nonce must be exactly {NONCE_SIZE} bytes")
    if not isinstance(ciphertext, (bytes, bytearray)):
        raise TypeError("ciphertext must be bytes")

    has_signature = sig_id != 0x0000
    if has_signature and signature is None:
        raise ValueError("signature is required when sig_id != 0x0000")
    if not has_signature and signature:
        raise ValueError("signature must be absent when sig_id == 0x0000")

    header = b"".join((
        MAGIC,
        bytes([VERSION]),
        _u16_be(kem_id),
        _u16_be(sig_id),
        bytes([flags]),
    ))
    # Invariant: the spec requires exactly 10 bytes of header.
    assert len(header) == HEADER_SIZE

    parts = [
        header,
        _u32_be(len(ct_kem)), bytes(ct_kem),
        _u32_be(len(sender_pub)), bytes(sender_pub),
    ]
    if has_signature:
        parts.append(_u32_be(len(signature)))
        parts.append(bytes(signature))
    parts.append(bytes(nonce))
    parts.append(_u32_be(len(ciphertext)))
    parts.append(bytes(ciphertext))

    return b"".join(parts)


def decode(blob: bytes) -> dict:
    """Parse a v1 blob. Raises InvalidMagic/InvalidVersion/InvalidFlags/MalformedBlob.

    Returns a dict with keys: version, kem_id, sig_id, flags, ct_kem, sender_pub,
    signature (None if sig_id==0), nonce, ciphertext, aad, consumed_bytes.
    """
    if not isinstance(blob, (bytes, bytearray)):
        raise TypeError("blob must be bytes")
    if len(blob) < HEADER_SIZE:
        raise MalformedBlob("blob too short for header")

    if blob[0:4] != MAGIC:
        raise InvalidMagic(bytes(blob[0:4]))

    version = blob[4]
    if version not in SUPPORTED_VERSIONS:
        raise InvalidVersion(version, SUPPORTED_VERSIONS)

    kem_id = int.from_bytes(blob[5:7], "big")
    sig_id = int.from_bytes(blob[7:9], "big")
    flags = blob[9]
    if flags != 0x00:
        raise InvalidFlags(flags)

    off = HEADER_SIZE

    def read_lp(label: str) -> bytes:
        nonlocal off
        if len(blob) < off + 4:
            raise MalformedBlob(f"truncated at {label} length")
        n = int.from_bytes(blob[off:off + 4], "big")
        off += 4
        if len(blob) < off + n:
            raise MalformedBlob(f"truncated at {label} body (need {n}, have {len(blob) - off})")
        out = bytes(blob[off:off + n])
        off += n
        return out

    ct_kem = read_lp("ct_kem")
    sender_pub = read_lp("sender_pub")

    signature = None
    if sig_id != 0x0000:
        signature = read_lp("signature")

    if len(blob) < off + NONCE_SIZE:
        raise MalformedBlob("truncated at nonce")
    nonce = bytes(blob[off:off + NONCE_SIZE])
    off += NONCE_SIZE

    ciphertext = read_lp("ciphertext")

    return {
        "version": version,
        "kem_id": kem_id,
        "sig_id": sig_id,
        "flags": flags,
        "ct_kem": ct_kem,
        "sender_pub": sender_pub,
        "signature": signature,
        "nonce": nonce,
        "ciphertext": ciphertext,
        "aad": bytes(blob[0:HEADER_SIZE]),
        "consumed_bytes": off,
    }


def build_aad(kem_id: int, sig_id: int, flags: int = 0x00, chunk_index: int = 0) -> bytes:
    """Build the AES-256-GCM AAD per spec: header(10) || chunk_index_be32."""
    if flags != 0x00:
        raise InvalidFlags(flags)
    return b"".join((
        MAGIC,
        bytes([VERSION]),
        _u16_be(kem_id),
        _u16_be(sig_id),
        bytes([flags]),
        _u32_be(chunk_index),
    ))


def is_v1(blob: bytes) -> bool:
    """Return True iff blob starts with the PQHB magic."""
    return isinstance(blob, (bytes, bytearray)) and len(blob) >= 4 and blob[0:4] == MAGIC

File: sdk-js/src/wire-format.js
// Paramant wire-format v1 encoder / decoder.
// Bit-exact against docs/wire-format-v1.md.

export const MAGIC = new Uint8Array([0x50, 0x51, 0x48, 0x42]); // 'PQHB'
export const VERSION_V1 = 0x01;
export const SUPPORTED_VERSIONS = [VERSION_V1];
export const HEADER_FIXED_SIZE = 10;
export const NONCE_SIZE = 12;

export class InvalidMagicError extends Error {
  constructor(got) { super(`invalid magic: ${got}`); this.name = 'InvalidMagicError'; }
}
export class InvalidVersionError extends Error {
  constructor(got) { super(`invalid version: ${got}`); this.name = 'InvalidVersionError'; }
}
export class MalformedBlobError extends Error {
  constructor(msg) { super(`malformed blob: ${msg}`); this.name = 'MalformedBlobError'; }
}
export class InvalidFlagsError extends Error {
  constructor(got) { super(`invalid flags: ${got}`); this.name = 'InvalidFlagsError'; }
}

function asU8(x, name) {
  if (x instanceof Uint8Array) return x;
  if (x && x.buffer instanceof ArrayBuffer) return new Uint8Array(x.buffer, x.byteOffset, x.byteLength);
  throw new TypeError(`${name} must be a Uint8Array`);
}

export function encode({ kemId, sigId, flags = 0x00, ctKem, senderPub, signature, nonce, ciphertext }) {
  if (!Number.isInteger(kemId) || kemId < 0 || kemId > 0xFFFF) throw new Error('kemId must be uint16');
  if (!Number.isInteger(sigId) || sigId < 0 || sigId > 0xFFFF) throw new Error('sigId must be uint16');
  if (flags !== 0x00) throw new InvalidFlagsError(flags);

  const ctKemU8 = asU8(ctKem, 'ctKem');
  const senderPubU8 = asU8(senderPub, 'senderPub');
  const nonceU8 = asU8(nonce, 'nonce');
  const ciphertextU8 = asU8(ciphertext, 'ciphertext');

  if (nonceU8.length !== NONCE_SIZE) throw new Error(`nonce must be ${NONCE_SIZE} bytes`);

  const hasSignature = sigId !== 0x0000;
  let signatureU8 = null;
  if (hasSignature) {
    if (!signature) throw new Error('signature required when sigId != 0x0000');
    signatureU8 = asU8(signature, 'signature');
  } else if (signature) {
    throw new Error('signature must be absent when sigId = 0x0000');
  }

  const total =
    HEADER_FIXED_SIZE +
    4 + ctKemU8.length +
    4 + senderPubU8.length +
    (hasSignature ? 4 + signatureU8.length : 0) +
    NONCE_SIZE +
    4 + ciphertextU8.length;

  const out = new Uint8Array(total);
  const dv = new DataView(out.buffer, out.byteOffset, out.byteLength);
  let o = 0;

  // Header
  out.set(MAGIC, o); o += 4;
  dv.setUint8(o, VERSION_V1); o += 1;
  dv.setUint16(o, kemId, false); o += 2;
  dv.setUint16(o, sigId, false); o += 2;
  dv.setUint8(o, flags); o += 1;

  // Key encapsulation
  dv.setUint32(o, ctKemU8.length, false); o += 4;
  out.set(ctKemU8, o); o += ctKemU8.length;
  dv.setUint32(o, senderPubU8.length, false); o += 4;
  out.set(senderPubU8, o); o += senderPubU8.length;

  // Signature (if signed)
  if (hasSignature) {
    dv.setUint32(o, signatureU8.length, false); o += 4;
    out.set(signatureU8, o); o += signatureU8.length;
  }

  // Payload
  out.set(nonceU8, o); o += NONCE_SIZE;
  dv.setUint32(o, ciphertextU8.length, false); o += 4;
  out.set(ciphertextU8, o); o += ciphertextU8.length;

  return out;
}

export function decode(blob) {
  const b = asU8(blob, 'blob');
  if (b.length < HEADER_FIXED_SIZE) throw new MalformedBlobError('too short for header');

  for (let i = 0; i < 4; i++) {
    if (b[i] !== MAGIC[i]) {
      const hex = [...b.slice(0, 4)].map(x => x.toString(16).padStart(2, '0')).join('');
      throw new InvalidMagicError(hex);
    }
  }

  const dv = new DataView(b.buffer, b.byteOffset, b.byteLength);
  const version = b[4];
  if (!SUPPORTED_VERSIONS.includes(version)) throw new InvalidVersionError(version);

  const kemId = dv.getUint16(5, false);
  const sigId = dv.getUint16(7, false);
  const flags = b[9];
  if (flags !== 0x00) throw new InvalidFlagsError(flags);

  let o = HEADER_FIXED_SIZE;

  if (b.length < o + 4) throw new MalformedBlobError('truncated at ctKem length');
  const ctKemLen = dv.getUint32(o, false); o += 4;
  if (b.length < o + ctKemLen) throw new MalformedBlobError('truncated at ctKem body');
  const ctKem = b.slice(o, o + ctKemLen); o += ctKemLen;

  if (b.length < o + 4) throw new MalformedBlobError('truncated at senderPub length');
  const senderPubLen = dv.getUint32(o, false); o += 4;
  if (b.length < o + senderPubLen) throw new MalformedBlobError('truncated at senderPub body');
  const senderPub = b.slice(o, o + senderPubLen); o += senderPubLen;

  let signature = null;
  if (sigId !== 0x0000) {
    if (b.length < o + 4) throw new MalformedBlobError('truncated at sig length');
    const sigLen = dv.getUint32(o, false); o += 4;
    if (b.length < o + sigLen) throw new MalformedBlobError('truncated at sig body');
    signature = b.slice(o, o + sigLen); o += sigLen;
  }

  if (b.length < o + NONCE_SIZE) throw new MalformedBlobError('truncated at nonce');
  const nonce = b.slice(o, o + NONCE_SIZE); o += NONCE_SIZE;

  if (b.length < o + 4) throw new MalformedBlobError('truncated at ciphertext length');
  const ctLen = dv.getUint32(o, false); o += 4;
  if (b.length < o + ctLen) throw new MalformedBlobError('truncated at ciphertext body');
  const ciphertext = b.slice(o, o + ctLen); o += ctLen;

  return {
    version, kemId, sigId, flags,
    ctKem, senderPub, signature, nonce, ciphertext,
    aad: b.slice(0, HEADER_FIXED_SIZE),
    consumedBytes: o,
  };
}

export function buildAAD({ kemId, sigId, flags = 0x00, chunkIndex = 0 }) {
  const buf = new Uint8Array(HEADER_FIXED_SIZE + 4);
  const dv = new DataView(buf.buffer);
  buf.set(MAGIC, 0);
  dv.setUint8(4, VERSION_V1);
  dv.setUint16(5, kemId, false);
  dv.setUint16(7, sigId, false);
  dv.setUint8(9, flags);
  dv.setUint32(HEADER_FIXED_SIZE, chunkIndex, false);
  return buf;
}

export function isV1(blob) {
  if (!(blob instanceof Uint8Array)) return false;
  if (blob.length < 4) return false;
  for (let i = 0; i < 4; i++) if (blob[i] !== MAGIC[i]) return false;
  return true;
}
